apps/labs/posts/narwhals-pycapsule.md

```diff
@@ -0,0 +1,199 @@
+
+## Python solution: Narwhals
+
+Narwhals is a lightweight, dependency-free, and extensible compatibility layer between
+dataframe libraries functions. To use it to write a dataframe-agnostic function,
+you need to follow three steps:
+
+- Call `narwhals.from_native` on the user's object.
+- Use the [Narwhals API](https://narwhals-dev.github.io/narwhals/api-reference/).
+- If you want to return to the user an object of the same kind that they started with,
+  call `to_native`.
+
+In this case, a Narwhals version of `agnostic_sum_i64_column` looks like this:
+
+```python
+import narwhals as nw
+from narwhals.typing import IntoFrame
+
+
+def agnostic_sum_i64_column_narwhals(df_native: IntoFrame, column_name: str) -> int:
+    lf = nw.from_native(df_native).lazy()
+    schema = lf.collect_schema()
+    if column_name not in schema:
+        msg = f"Column '{column_name}' not found, available columns are: {schema.names()}."
+        raise ValueError(msg)
+    if (dtype := schema[column_name]) != nw.Int64:
+        msg = f"Column '{column_name}' is of type {dtype}, expected Int64"
+        raise TypeError(msg)
+    df = lf.collect()
+    return df[column_name].sum()
+```
+
+Like above, we can now pass different inputs and get the same result:
+```python
+>>> agnostic_sum_i64_column_narwhals(df_pl, column_name="a")
+6
+>>> agnostic_sum_i64_column_narwhals(df_pd, column_name="a")
+6
+>>> agnostic_sum_i64_column_narwhals(df_pa, column_name="a")
+6
+>>> agnostic_sum_i64_column_narwhals(rel, column_name="a")
+6
+```
+
+So long as you stick to the Narwhals API, your code will keep working with all major dataframe libraries,
+as well as with new ones which may appear in the future!
+
+## Narwhals vs PyCapsule Interface: when to use one over the other, and when to use them together
+
+If the Narwhals API is extensive enough for your use-case, then this is arguably
+a simpler and easier solution than writing your own Rust function. On the other hand,
+if you write a Rust function using the PyCapsule Interface, then you have complete
+control over the data. So, when should you use which?
+
+Let's cover some scenarios:
+
+- If you want your dataframe logic to stay completely lazy where possible: use Narwhals.
+  The PyCapsule Interface requires you to materialise the data into memory immediately,
+  whereas Narwhals has both eager and lazy APIs.
+- If you want complete control over your data: use the
+  PyCapsule Interface. If you have the necessary Rust / C skills, there should be no limit
+  to how complex and bespoke you make your data processing.
+- If you want to keep your library logic to pure-Python and without heavy dependencies so
+  it's easy to maintain and install: use Narwhals. Packaging a pure-Python project is very
+  easy, especially compared with if you need to get Rust or C in there.
+- If you want to do part of your processing in Python, and part of it in Rust - **use both**!
```

Comment:
If you cared to cater to non-Rust developers, you might also want to mention nanoarrow. I've also done a youtube video on building Python extensions using C++ - https://www.youtube.com/watch?v=EhUnmXPjTy8&list=PLyPvk-ejh-138onAgM7sVcpzu1HLp-Uiv

Not trying to promote my own content here - just offering in case you wanted to expand the scope of your blog post (which may or may not be worth it)

Reply:
sounds good, happy to link it, just not sure how to phrase it - any suggestions?